### DISCUSSION TITLE
String representation of floats in queries

### DISCUSSION THREAD

**Discussion by paradiddle-luuk** [duckdb/duckdb#10788] [General]
# Introduction
Hello, I use DuckDB for vector embeddings, but my question relates in general to FLOAT lists and arrays in DuckDB and in PGVector(Postgres).

Both systems want me to represent floating point values as a (comma separated) string interpolated into SQL queries that insert values into the Array/list column. This makes me feel a bit sad as we are talking about single or double precision floats that are forced into string representations just to be stored as floating points in the database.

### DuckDB with Python f-string
Here I am avoiding certain type errors by using array_value explicitely:
```
query.query_embedding = [1/3, math.sqrt(2), 1/6]
con.execute(
    f"SELECT *, array_cosine_similarity(embedding, array_value({",".join(map(str, query.query_embedding))})) AS score ...;"
)
```
or straight out writing string representations in an insert statement

```
con.execute("INSERT INTO embeddings VALUES (1, [1.33333, 2.0, 3.0]);")
```

### Postgres (PGVector) with percent syntax

```
embedding = np.array([1.1, 1/3, np.sqrt(2)])

conn.execute(
    "INSERT INTO embeddings (id, embedding) VALUES (%s, %s)", (1, embedding)
)
```
## Why is there no way (bulk) insert byte representations of floats?
I wonder what is forcing me to go through a string representation in these queries. Are we having floating point inaccuracies as a consequence of this?

I would like there to be a more natural way of inserting floating point content into DuckDB and databases in general.

**Comment by szarnyasg** [comment]
Hello @luuk-codebeez,

1) In Python, DuckDB can directly read pandas dataframes, so one course of action would be to read the floating point values from a pandas data frame: https://duckdb.org/docs/guides/python/sql_on_pandas. Would this work for your use case?

2) You can use PyArrow to pass floats, https://duckdb.org/docs/guides/python/sql_on_arrow

3) There is an appender in several APIs (C, C++, Go, Java, and Rust – but not in Python), which allows bulk inserting data: https://duckdb.org/docs/data/appender

**Reply by paradiddle-luuk** [reply]
Great reply, I will play around with these and see if I can improve some things in the llama-index package.
